wasmtime-environ/src/func_environ.rs

```diff
@@ -107,106 +126,102 @@ impl<'module_environment> FuncEnvironment<'module_environment> {
+                call_conv: self.target_config.default_call_conv,
+            })
+        });
+        self.memory32_size_sig = Some(sig);
+        sig
     }
 
-    /// Return the memory.size function to call for the given index, along with the
-    /// translated index value to pass to it.
+    /// Return the memory.size function signature to call for the given index, along with the
+    /// translated index value to pass to it and its index in `VMBuiltinFunctionsArray`.
     fn get_memory_size_func(
         &mut self,
         func: &mut Function,
         index: MemoryIndex,
-    ) -> (FuncRef, usize) {
+    ) -> (ir::SigRef, usize, BuiltinFunctionIndex) {
         if self.module.is_imported_memory(index) {
-            let extfunc = self.imported_memory32_size_extfunc.unwrap_or_else(|| {
-                let sig_ref = self.get_memory32_size_sig(func);
-                func.import_function(ExtFuncData {
-                    name: get_imported_memory32_size_name(),
-                    signature: sig_ref,
-                    // We currently allocate all code segments independently, so nothing
-                    // is colocated.
-                    colocated: false,
-                })
-            });
-            self.imported_memory32_size_extfunc = Some(extfunc);
-            (extfunc, index.index())
+            (
+                self.get_memory32_size_sig(func),
+                index.index(),
+                BuiltinFunctionIndex::get_imported_memory32_size_index(),
+            )
         } else {
-            let extfunc = self.memory32_size_extfunc.unwrap_or_else(|| {
-                let sig_ref = self.get_memory32_size_sig(func);
-                func.import_function(ExtFuncData {
-                    name: get_memory32_size_name(),
-                    signature: sig_ref,
-                    // We currently allocate all code segments independently, so nothing
-                    // is colocated.
-                    colocated: false,
-                })
-            });
-            self.memory32_size_extfunc = Some(extfunc);
             (
-                extfunc,
+                self.get_memory32_size_sig(func),
                 self.module.defined_memory_index(index).unwrap().index(),
+                BuiltinFunctionIndex::get_memory32_size_index(),
             )
         }
     }
+
+    /// Translates load of builtin function and returns a pair of values `vmctx`
+    /// and address of the loaded function.
+    fn translate_load_builtin_function_address(
+        &mut self,
+        pos: &mut FuncCursor<'_>,
+        callee_func_idx: BuiltinFunctionIndex,
+    ) -> (ir::Value, ir::Value) {
+        // We use an indirect call so that we don't have to patch the code at runtime.
+        let pointer_type = self.pointer_type();
+        let vmctx = self.vmctx(&mut pos.func);
+        let base = pos.ins().global_value(pointer_type, vmctx);
+
+        let mem_flags = ir::MemFlags::trusted();
```

Comment:
We can also call `set_readonly()` on the `mem_flags`, which indicates that the contents of the memory doesn't change at runtime, which allows the optimizer to eliminate redundant copies of this load if there are eg. multiple `memory.size` instructions.

Reply:
Thanks! I'll fix it. I read the translation of direct call, and missed that detail (it's in the translation of indirect call).